File: examples/bun-app/src/routes/services.ts
import { Hono } from 'hono';
import { eq } from 'drizzle-orm';
import { db } from '../db/client';
import { serviceRequests } from '../db/schema';
import { FfiLayer, TaskerClient } from '@tasker-systems/tasker';

export const servicesRoute = new Hono();

/**
 * POST /services/requests
 *
 * Creates a new service request record, then creates a Tasker task to orchestrate
 * the microservices user registration workflow (5 steps with diamond pattern).
 */
servicesRoute.post('/', async (c) => {
  const body = await c.req.json();
  const { username, email, plan, metadata } = body;

  if (!username || !email) {
    return c.json({ error: 'username and email are required' }, 400);
  }

  // Create domain record
  const [request] = await db
    .insert(serviceRequests)
    .values({
      username,
      email,
      plan: plan || 'free',
      metadata: metadata || {},
      status: 'pending',
    })
    .returning();

  // Create Tasker task for user registration
  let taskUuid: string | null = null;
  try {
    const ffiLayer = new FfiLayer();
    await ffiLayer.load();
    const client = new TaskerClient(ffiLayer);

    const task = client.createTask({
      name: 'user_registration',
      context: {
        request_id: request.id,
        username,
        email,
        plan: plan || 'free',
        metadata: metadata || {},
      },
      initiator: 'bun-app',
      sourceSystem: 'example-bun-app',
      reason: `Register user: ${username}`,
      tags: ['microservices', 'registration'],
      idempotencyKey: `registration-${request.id}`,
    });

    taskUuid = task.task_uuid;

    await db
      .update(serviceRequests)
      .set({ taskUuid, status: 'processing', updatedAt: new Date() })
      .where(eq(serviceRequests.id, request.id));
  } catch (error) {
    console.error('Failed to create Tasker task for service request:', error);
  }

  return c.json(
    {
      id: request.id,
      username: request.username,
      email: request.email,
      plan: request.plan,
      status: taskUuid ? 'processing' : 'pending',
      task_uuid: taskUuid,
      created_at: request.createdAt,
    },
    201,
  );
});

/**
 * GET /services/requests/:id
 *
 * Loads the service request record and, if a task_uuid exists, fetches the
 * current task status from Tasker for a combined view.
 */
servicesRoute.get('/:id', async (c) => {
  const id = parseInt(c.req.param('id'), 10);

  const request = await db.query.serviceRequests.findFirst({
    where: eq(serviceRequests.id, id),
  });

  if (!request) {
    return c.json({ error: 'Service request not found' }, 404);
  }

  let taskStatus = null;
  if (request.taskUuid) {
    try {
      const ffiLayer = new FfiLayer();
      await ffiLayer.load();
      const client = new TaskerClient(ffiLayer);
      taskStatus = client.getTask(request.taskUuid);
    } catch (error) {
      console.error('Failed to fetch task status:', error);
    }
  }

  return c.json({
    id: request.id,
    username: request.username,
    email: request.email,
    plan: request.plan,
    metadata: request.metadata,
    status: request.status,
    task_uuid: request.taskUuid,
    task_status: taskStatus,
    created_at: request.createdAt,
    updated_at: request.updatedAt,
  });
});

File: examples/fastapi-app/app/routes/services.py
"""Microservices user registration routes demonstrating diamond dependency pattern.

POST /services/requests/ - Create a service request and kick off the 5-step diamond workflow
GET  /services/requests/{id} - Get service request details with current task status
"""

from __future__ import annotations

import logging
import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import ServiceRequest
from app.schemas import CreateServiceRequest, ServiceRequestResponse

logger = logging.getLogger(__name__)
router = APIRouter()


@router.post("/requests/", response_model=ServiceRequestResponse, status_code=201)
async def create_service_request(
    request: CreateServiceRequest,
    db: AsyncSession = Depends(get_db),
) -> ServiceRequestResponse:
    """Create a user registration request and start the diamond workflow.

    Creates a domain record then creates a Tasker task for the 5-step pipeline:
    CreateUser -> (SetupBilling || InitPreferences) -> SendWelcome -> UpdateStatus.
    """
    from tasker_core.client import TaskerClient

    svc_request = ServiceRequest(
        user_id=request.user_id,
        request_type=request.request_type,
        status="pending",
    )
    db.add(svc_request)
    await db.flush()

    client = TaskerClient(initiator="fastapi-example", source_system="fastapi-example")
    try:
        task_response = client.create_task(
            "user_registration",
            namespace="microservices",
            context={
                "request_id": svc_request.id,
                "user_id": request.user_id,
                "email": request.email,
                "full_name": request.full_name,
                "plan": request.plan,
            },
            reason="User registration",
        )
        task_uuid = uuid.UUID(task_response.task_uuid)
        svc_request.task_uuid = task_uuid
        svc_request.status = "processing"
        logger.info(
            "Created task %s for service request %d", task_uuid, svc_request.id
        )
    except Exception:
        logger.exception(
            "Failed to create tasker task for service request %d", svc_request.id
        )
        svc_request.status = "task_creation_failed"

    await db.commit()
    await db.refresh(svc_request)

    return ServiceRequestResponse(
        id=svc_request.id,
        user_id=svc_request.user_id,
        request_type=svc_request.request_type,
        status=svc_request.status,
        result=svc_request.result,
        task_uuid=svc_request.task_uuid,
        created_at=svc_request.created_at,
        updated_at=svc_request.updated_at,
    )


@router.get("/requests/{request_id}", response_model=ServiceRequestResponse)
async def get_service_request(
    request_id: int,
    db: AsyncSession = Depends(get_db),
) -> ServiceRequestResponse:
    """Get service request details with current workflow task status."""
    from tasker_core.client import TaskerClient

    result = await db.execute(
        select(ServiceRequest).where(ServiceRequest.id == request_id)
    )
    svc_request = result.scalar_one_or_none()
    if svc_request is None:
        raise HTTPException(status_code=404, detail="Service request not found")

    task_status = None
    if svc_request.task_uuid:
        try:
            client = TaskerClient(initiator="fastapi-example", source_system="fastapi-example")
            task_status = client.get_task(str(svc_request.task_uuid))
        except Exception:
            logger.exception(
                "Failed to fetch task status for service request %d", svc_request.id
            )

    return ServiceRequestResponse(
        id=svc_request.id,
        user_id=svc_request.user_id,
        request_type=svc_request.request_type,
        status=svc_request.status,
        result=svc_request.result,
        task_uuid=svc_request.task_uuid,
        created_at=svc_request.created_at,
        updated_at=svc_request.updated_at,
        task_status=task_status,
    )
